## **Predict whether income exceeds $50K/yr based on census data. Also known as "Census Income" dataset.**

#### Approach 1

In [68]:
#Import Libraries
import pandas as pd

In [69]:
# Import data from local system
file_path = r'C:\NumpyNinja\ML classes -Project\ML Assignments\adult\adult.data'

df = pd.read_csv(file_path, delimiter=",", names=["age","workclass","fnlwgt","education","education-num","marital-status","occupation","relationship","race","sex","capital-gain","capital-loss","hours-per-week","native-country","income"])

df

age          workclass  fnlwgt    education  education-num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

#### **The number of rows here are 32561 and it is not matching with the number of instances given in dataset link. We decided to move forward with approach 2**

#### Approach 2

In [2]:
#Installing ucimlrepo package
!pip install ucimlrepo

In [22]:
#Import Libraries
import pandas as pd
import numpy as np
from ucimlrepo import fetch_ucirepo, list_available_datasets

In [4]:
#List of UCI repositories with ID
list_available_datasets()

-------------------------------------
The following datasets are available:
-------------------------------------
Dataset Name                                                                            ID    
------------                                                                            --    
Abalone                                                                                 1     
Adult                                                                                   2     
Annealing                                                                               3     
Audiology (Standardized)                                                                8     
Auto MPG                                                                                9     
Automobile                                                                              10    
Balance Scale                                                                           12    
Balloons                       

In [5]:
# Fetch the 'adult' dataset from UCI repository where id for dataset is 2
adultDataset = fetch_ucirepo(id=2)

In [6]:
#Creating dataframe by joining features and target
Xdata = pd.DataFrame(adultDataset.data.features)
Ydata = pd.DataFrame(adultDataset.data.targets, columns=['income'])
adultDataFrame = pd.concat([Xdata,Ydata], axis = 1)
adultDataFrame

age         workclass  fnlwgt  education  education-num  \
0       39         State-gov   77516  Bachelors             13   
1       50  Self-emp-not-inc   83311  Bachelors             13   
2       38           Private  215646    HS-grad              9   
3       53           Private  234721       11th              7   
4       28           Private  338409  Bachelors             13   
...    ...               ...     ...        ...            ...   
48837   39           Private  215419  Bachelors             13   
48838   64               NaN  321403    HS-grad              9   
48839   38           Private  374983  Bachelors             13   
48840   44           Private   83891  Bachelors             13   
48841   35      Self-emp-inc  182148  Bachelors             13   

           marital-status         occupation    relationship  \
0           Never-married       Adm-clerical   Not-in-family   
1      Married-civ-spouse    Exec-managerial         Husband   
2                Divorced  Handlers-cleaners   Not-in-family   
3      Married-civ-spouse  Handlers-cleaners         Husband   
4      Married-civ-spouse     Prof-specialty            Wife   
...                   ...                ...             ...   
48837            Divorced     Prof-specialty   Not-in-family   
48838             Widowed                NaN  Other-relative   
48839  Married-civ-spouse     Prof-specialty         Husband   
48840            Divorced       Adm-clerical       Own-child   
48841  Married-civ-spouse    Exec-managerial         Husband   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
0                   White    Male          2174             0              40   
1                   White    Male             0             0              13   
2                   White    Male             0             0              40   
3                   Black    Male             0             0              40   
4                   Black  Female             0             0              40   
...                   ...     ...           ...           ...             ...   
48837               White  Female             0             0              36   
48838               Black    Male             0             0              40   
48839               White    Male             0             0              50   
48840  Asian-Pac-Islander    Male          5455             0              40   
48841               White    Male             0             0              60   

      native-country  income  
0      United-States   <=50K  
1      United-States   <=50K  
2      United-States   <=50K  
3      United-States   <=50K  
4               Cuba   <=50K  
...              ...     ...  
48837  United-States  <=50K.  
48838  United-States  <=50K.  
48839  United-States  <=50K.  
48840  United-States  <=50K.  
48841  United-States   >50K.  

[48842 rows x 15 columns]

In [7]:
# Displaying first 10 rows of dataset
adultDataFrame.head(10)

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   
5   37           Private  284582    Masters             14   
6   49           Private  160187        9th              5   
7   52  Self-emp-not-inc  209642    HS-grad              9   
8   31           Private   45781    Masters             14   
9   42           Private  159449  Bachelors             13   

          marital-status         occupation   relationship   race     sex  \
0          Never-married       Adm-clerical  Not-in-family  White    Male   
1     Married-civ-spouse    Exec-managerial        Husband  White    Male   
2               Divorced  Handlers-cleaners  Not-in-family  White    Male   
3     Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4     Married-civ-spouse     Prof-specialty           Wife  Black  Female   
5     Married-civ-spouse    Exec-managerial           Wife  White  Female   
6  Married-spouse-absent      Other-service  Not-in-family  Black  Female   
7     Married-civ-spouse    Exec-managerial        Husband  White    Male   
8          Never-married     Prof-specialty  Not-in-family  White  Female   
9     Married-civ-spouse    Exec-managerial        Husband  White    Male   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K  
5             0             0              40  United-States  <=50K  
6             0             0              16        Jamaica  <=50K  
7             0             0              45  United-States   >50K  
8         14084             0              50  United-States   >50K  
9          5178             0              40  United-States   >50K

In [8]:
# Displaying 5 rows using sample function
adultDataFrame.sample(n=5,random_state = 42)

age  workclass  fnlwgt     education  education-num marital-status  \
7762    18    Private  423024       HS-grad              9  Never-married   
23881   17    Private  178953          12th              8  Never-married   
30507   25  Local-gov  348986       HS-grad              9  Never-married   
28911   20    Private  218215  Some-college             10  Never-married   
19484   47    Private  244025       HS-grad              9  Never-married   

              occupation    relationship                race     sex  \
7762       Other-service   Not-in-family               White    Male   
23881              Sales       Own-child               White  Female   
30507  Handlers-cleaners  Other-relative               Black    Male   
28911              Sales       Own-child               White  Female   
19484  Machine-op-inspct       Unmarried  Amer-Indian-Eskimo    Male   

       capital-gain  capital-loss  hours-per-week native-country income  
7762              0             0              20  United-States  <=50K  
23881             0             0              20  United-States  <=50K  
30507             0             0              40  United-States  <=50K  
28911             0             0              30  United-States  <=50K  
19484             0             0              56    Puerto-Rico  <=50K

In [9]:
# Examining columns using info
adultDataFrame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             48842 non-null  int64 
 1   workclass       47879 non-null  object
 2   fnlwgt          48842 non-null  int64 
 3   education       48842 non-null  object
 4   education-num   48842 non-null  int64 
 5   marital-status  48842 non-null  object
 6   occupation      47876 non-null  object
 7   relationship    48842 non-null  object
 8   race            48842 non-null  object
 9   sex             48842 non-null  object
 10  capital-gain    48842 non-null  int64 
 11  capital-loss    48842 non-null  int64 
 12  hours-per-week  48842 non-null  int64 
 13  native-country  48568 non-null  object
 14  income          48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


In [10]:
# Checking for missing values using isnull
adultDataFrame.isnull().sum()

age                 0
workclass         963
fnlwgt              0
education           0
education-num       0
marital-status      0
occupation        966
relationship        0
race                0
sex                 0
capital-gain        0
capital-loss        0
hours-per-week      0
native-country    274
income              0
dtype: int64

In [11]:
# Summary statistics for numerical columns using describe 
adultDataFrame.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  48842.000000  4.884200e+04   48842.000000  48842.000000  48842.000000   
mean      38.643585  1.896641e+05      10.078089   1079.067626     87.502314   
std       13.710510  1.056040e+05       2.570973   7452.019058    403.004552   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.175505e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.781445e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.376420e+05      12.000000      0.000000      0.000000   
max       90.000000  1.490400e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    48842.000000  
mean        40.422382  
std         12.391444  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

In [12]:
# For checking summary statistics for categorical columns
adultDataFrame.describe(include='object')

workclass education      marital-status      occupation relationship  \
count      47879     48842               48842           47876        48842   
unique         9        16                   7              15            6   
top      Private   HS-grad  Married-civ-spouse  Prof-specialty      Husband   
freq       33906     15784               22379            6172        19716   

         race    sex native-country income  
count   48842  48842          48568  48842  
unique      5      2             42      4  
top     White   Male  United-States  <=50K  
freq    41762  32650          43832  24720

**Data Cleaning:
Handle missing values by either removing rows or filling them with appropriate values (mean, median, mode, etc.). Correct data types if needed. Address any outliers or anomalies**

In [13]:
#checking for null values in the columns
adultDataFrame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             48842 non-null  int64 
 1   workclass       47879 non-null  object
 2   fnlwgt          48842 non-null  int64 
 3   education       48842 non-null  object
 4   education-num   48842 non-null  int64 
 5   marital-status  48842 non-null  object
 6   occupation      47876 non-null  object
 7   relationship    48842 non-null  object
 8   race            48842 non-null  object
 9   sex             48842 non-null  object
 10  capital-gain    48842 non-null  int64 
 11  capital-loss    48842 non-null  int64 
 12  hours-per-week  48842 non-null  int64 
 13  native-country  48568 non-null  object
 14  income          48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


In [14]:
adultDataFrame.isnull().sum()

age                 0
workclass         963
fnlwgt              0
education           0
education-num       0
marital-status      0
occupation        966
relationship        0
race                0
sex                 0
capital-gain        0
capital-loss        0
hours-per-week      0
native-country    274
income              0
dtype: int64

In [15]:
(adultDataFrame == '?').sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loss         0
hours-per-week       0
native-country     583
income               0
dtype: int64

In [27]:

adult_df_cleaned = adultDataFrame.replace('?', np.nan)
adult_df_cleaned.isnull().sum()

age                  0
workclass         2799
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        2809
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loss         0
hours-per-week       0
native-country     857
income               0
dtype: int64

In [32]:
adult_df_cleaned['workclass'].unique()

array(['State-gov', 'Self-emp-not-inc', 'Private', 'Federal-gov',
       'Local-gov', nan, 'Self-emp-inc', 'Without-pay', 'Never-worked'],
      dtype=object)

In [28]:
adult_df_cleaned['workclass']

2799

In [29]:
for column in adult_df_cleaned.columns:
    print(f"--------------------{column.title()}-------------------------")
    print(adult_df_cleaned[column].value_counts())

--------------------Age-------------------------
age
36    1348
35    1337
33    1335
23    1329
31    1325
      ... 
88       6
85       5
87       3
89       2
86       1
Name: count, Length: 74, dtype: int64
--------------------Workclass-------------------------
workclass
Private             33906
Self-emp-not-inc     3862
Local-gov            3136
State-gov            1981
Self-emp-inc         1695
Federal-gov          1432
Without-pay            21
Never-worked           10
Name: count, dtype: int64
--------------------Fnlwgt-------------------------
fnlwgt
203488    21
120277    19
190290    19
125892    18
126569    18
          ..
286983     1
185942     1
234220     1
214706     1
350977     1
Name: count, Length: 28523, dtype: int64
--------------------Education-------------------------
education
HS-grad         15784
Some-college    10878
Bachelors        8025
Masters          2657
Assoc-voc        2061
11th             1812
Assoc-acdm       1601
10th             1389
7th-8

In [31]:
row_withnullval = adult_df_cleaned[adult_df_cleaned['workclass'].isnull()]
row_withnullval.head(20)

age workclass  fnlwgt     education  education-num  \
27    54       NaN  180211  Some-college             10   
61    32       NaN  293936       7th-8th              4   
69    25       NaN  200681  Some-college             10   
77    67       NaN  212759          10th              6   
106   17       NaN  304873          10th              6   
128   35       NaN  129305       HS-grad              9   
149   43       NaN  174662  Some-college             10   
154   52       NaN  252903       HS-grad              9   
160   68       NaN   38317       1st-4th              2   
187   53       NaN  135105     Bachelors             13   
201   19       NaN  170653       HS-grad              9   
221   64       NaN  187656       1st-4th              2   
226   60       NaN   24215          10th              6   
243   20       NaN  119156  Some-college             10   
266   19       NaN  218956  Some-college             10   
297   39       NaN  157443       Masters             14   
312   19       NaN  860348  Some-college             10   
326   23       NaN  211601     Assoc-voc             11   
346   67       NaN   36135          11th              7   
347   30       NaN  151989     Assoc-voc             11   

            marital-status occupation   relationship                race  \
27      Married-civ-spouse        NaN        Husband  Asian-Pac-Islander   
61   Married-spouse-absent        NaN  Not-in-family               White   
69           Never-married        NaN      Own-child               White   
77      Married-civ-spouse        NaN        Husband               White   
106          Never-married        NaN      Own-child               White   
128     Married-civ-spouse        NaN        Husband               White   
149               Divorced        NaN  Not-in-family               White   
154               Divorced        NaN  Not-in-family               White   
160               Divorced        NaN  Not-in-family               White   
187               Divorced        NaN  Not-in-family               White   
201          Never-married        NaN      Own-child               White   
221               Divorced        NaN  Not-in-family               White   
226               Divorced        NaN  Not-in-family  Amer-Indian-Eskimo   
243          Never-married        NaN      Own-child               White   
266          Never-married        NaN      Own-child               White   
297     Married-civ-spouse        NaN           Wife  Asian-Pac-Islander   
312          Never-married        NaN      Own-child               Black   
326          Never-married        NaN      Own-child               Black   
346     Married-civ-spouse        NaN        Husband               White   
347               Divorced        NaN      Unmarried               White   

        sex  capital-gain  capital-loss  hours-per-week native-country income  
27     Male             0             0              60          South   >50K  
61     Male             0             0              40            NaN  <=50K  
69     Male             0             0              40  United-States  <=50K  
77     Male             0             0               2  United-States  <=50K  
106  Female         34095             0              32  United-States  <=50K  
128    Male             0             0              40  United-States  <=50K  
149  Female             0             0              40  United-States  <=50K  
154    Male             0             0              45  United-States   >50K  
160  Female             0             0              20  United-States  <=50K  
187  Female             0             0              50  United-States  <=50K  
201    Male             0             0              40          Italy  <=50K  
221    Male             0             0              40  United-States  <=50K  
226  Female             0             0              10  United-States  <=50K  
243    Male             0             0         

In [37]:
adult_df_mode = adult_df_cleaned['workclass'].fillna(adult_df_cleaned['workclass'].mode().iloc[0])
adult_df_mode.value_counts()


workclass
Private             36705
Self-emp-not-inc     3862
Local-gov            3136
State-gov            1981
Self-emp-inc         1695
Federal-gov          1432
Without-pay            21
Never-worked           10
Name: count, dtype: int64

In [38]:
adult_df_cleaned['workclass'].value_counts()

workclass
Private             33906
Self-emp-not-inc     3862
Local-gov            3136
State-gov            1981
Self-emp-inc         1695
Federal-gov          1432
Without-pay            21
Never-worked           10
Name: count, dtype: int64

In [43]:
adult_df_cleaned['income'].value_counts()

income
<=50K     24720
<=50K.    12435
>50K       7841
>50K.      3846
Name: count, dtype: int64

In [47]:
adult_df_cleaned['income'].replace('<=50K.', '<=50K',inplace = True)
adult_df_cleaned['income'].replace('>50K.', '>50K',inplace = True)
adult_df_cleaned['income'].value_counts()

/var/folders/ln/pk1mgs_514n7ydhny2twzb680000gn/T/ipykernel_32365/1138969651.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  adult_df_cleaned['income'].replace('<=50K.', '<=50K',inplace = True)
/var/folders/ln/pk1mgs_514n7ydhny2twzb680000gn/T/ipykernel_32365/1138969651.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are

income
<=50K    37155
>50K     11687
Name: count, dtype: int64

In [50]:
adult_df_cleaned['occupation'].value_counts()

occupation
Prof-specialty       6172
Craft-repair         6112
Exec-managerial      6086
Adm-clerical         5611
Sales                5504
Other-service        4923
Machine-op-inspct    3022
Transport-moving     2355
Handlers-cleaners    2072
Farming-fishing      1490
Tech-support         1446
Protective-serv       983
Priv-house-serv       242
Armed-Forces           15
Name: count, dtype: int64